add dynamic shapes support for subclasses that override size/stride

bdhirsh opened the pull request:
This is mostly a minor fix on top of @soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #107917
* #106404
* #107926
* __->__ #107916
* #107915

bdhirsh committed:
[not working] add dynamic shapes support for subclasses that override size/stride

[ghstack-poisoned]

bdhirsh commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -720,9 +720,8 @@ static PyObject* THPVariable_make_wrapper_subclass(
-          false,
-          "Setting sizes_strides_policy isn't supported for this overload")
+      tensor.unsafeGetTensorImpl()->set_python_custom_sizes_strides(
+          parseSizesStridesPolicyArgument(*sizes_strides_policy));
```
I actually don't think I need this, double checking now

soulitzer commented on torch/csrc/PyInterpreter.cpp:
```diff
@@ -565,37 +576,21 @@ c10::IntArrayRef ConcretePyInterpreterVTable::strides(
+    ptr[idx] = py::cast<int64_t>(*it);
+  }
+  set_tensor_attr_with_capsule(
+      const_cast<c10::TensorImpl*>(self), capsule, "_sizes_capsule");
```
```suggestion
      const_cast<c10::TensorImpl*>(self), capsule, "_strides_capsule");
```

bdhirsh replied:
whoops- thanks!

bdhirsh committed:
Update on "add dynamic shapes support for subclasses that override size/stride"

This is mostly a minor fix on top of soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.




[ghstack-poisoned]

ezyang commented on test/test_python_dispatch.py:
```diff
@@ -2028,6 +2028,53 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
+            x_wrapper = CustomSizeDynamicShapesTensor(x)
+            return x_wrapper.size(), x_wrapper.stride()
+        fx_g = make_fx(trace_fn, tracing_mode="symbolic")(x)
+        # This is broken
```
Is it? Seems like it does what the torch dispatch said it should do?

bdhirsh replied:
**was** broken 🙂. Forgot to remove (during local testing)

soulitzer reviewed (approved):
Thanks!

soulitzer commented on test/test_python_dispatch.py:
```diff
@@ -2028,6 +2028,53 @@ def __torch_dispatch__(cls, func, types, args, kwargs):
+                return NotImplemented
+
+        x = torch.ones(2, 2)
+        def trace_fn(x):
```
lint

bdhirsh committed:
Update on "add dynamic shapes support for subclasses that override size/stride"

This is mostly a minor fix on top of soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.




[ghstack-poisoned]

bdhirsh committed:
Update on "add dynamic shapes support for subclasses that override size/stride"

This is mostly a minor fix on top of soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.




[ghstack-poisoned]

bdhirsh committed:
Update on "add dynamic shapes support for subclasses that override size/stride"

This is mostly a minor fix on top of soulitzer's PR https://github.com/pytorch/pytorch/pull/107839.

(1) `strides` wasn't going through the new `set_tensor_attr_with_capsule` flow
(2) The dynamic shapes overload for `_make_wrapper_subclass` currently errors when you try to use custom sizes - I removed the error
(3) added a test

I need this later because I'm adding a `__torch_dispatch__` `FunctionalTensor` wrapper subclass, that needs to support dynamic shapes, and also plumb metadata calls to its inner tensor later.




[ghstack-poisoned]
